### PR TITLE
SNS  - ADC multiplexer

### PR DESCRIPTION
**Pull request by licornes-fluos**
This is a new branch because the previous one was in the wrong place.

### CODE REVIEW

File: lib/sensors/adc_mux.hpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+#include <array>
+#include <cstdint>
+#include <memory>
+#include <optional>
+
+#include <core/logger.hpp>
+#include <core/types.hpp>
+#include <io/i2c.hpp>
+
+namespace hyped::adc {
```
Comment (DylanCavers) [inline lib/sensors/adc_mux.hpp:14]: should this be in `hyped::sensors`?

**Review by DylanCavers — COMMENTED** [review]
lgtm

File: lib/sensors/adc_mux.hpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+std::optional<std::array<T, N>> AdcMux<T, N>::readData()
+{
+  std::array<T, N> mux_data{};
+  std::uint8_t num_unusable_sensors = 0;
+  for (std::uint8_t i = 0; i < N; ++i) {
+    const auto &sensor            = sensors.at(i);
+    const std::uint8_t sensor_reg = sensor->getRegister();
+    const auto sensor_data        = sensor->read(sensor_reg) if (!sensor_data)
+    {
+      logger_.log(core::LogLevel::kFatal, "Failed to get i2c mux data from channel %d", channel);
```
Comment (TomLonergan03) [inline lib/sensors/adc_mux.hpp:68]: `adc mus data surely lol`

**Comment by TomLonergan03** [comment]
None of this is necessary, instead we just have a header file which contains `kAdcMuxAddress` and `kAdcChannel[1..n]`. Then for an ADC mux sensor it just takes a address and register for the register that channel is stored in.

e.g.
`adc_mux.hpp`
```c++
// boilerplate
constexpr kAdcMuxAddress = 0x13;
constexpr kAdcMuxChannel1 = 0x24;
...
constexpr kAdcMuxChannelN = 0x24+N;
// more boilerplate
```

`an_adc_mux_sensor.cpp`
```c++
#include adc_mux.hpp
#include hardware_i2c.hpp

// stuff

read() {
    i2c_.readByte(kAdcMuxAddress, kAdcMuxChannel1_);
}
`

**Pull request closed without merging**